version: '3.7'

services:
  web:
    environment:
      - SQLALCHEMY_DATABASE_URI=mysql+pymysql://root:root@mysql/fyp
      - SQLALCHEMY_TRACK_MODIFICATIONS=False
      - CELERY_BROKER_URL=redis://redis:6379/0
      - CELERY_RESULT_BACKEND=redis://redis:6379/0
    build: .
    ports:
      - "5000:5000"
    volumes:
      - .:/myapp
    depends_on:
      - mysql
      - redis

  celery:
    build: .
    environment:
      - CELERY_BROKER_URL=redis://redis:6379/0
      - CELERY_RESULT_BACKEND=redis://redis:6379/0
      - SQLALCHEMY_DATABASE_URI=mysql+pymysql://root:root@mysql/fyp
      - SQLALCHEMY_TRACK_MODIFICATIONS=False
    command: python3 -m celery -A run.celery worker --loglevel=info
    volumes:
      - .:/myapp
    depends_on:
      - web
      - redis

  mysql:
    image: mysql:5.7
    environment:
      MYSQL_ROOT_PASSWORD: root
      MYSQL_DATABASE: fyp
      MYSQL_USER: myuser
      MYSQL_PASSWORD: mypassword
      SQLALCHEMY_DATABASE_URI: "mysql+pymysql://root:root@mysql/fyp"
      SQLALCHEMY_TRACK_MODIFICATIONS: "False"
    ports:
      - "3307:3306"
    volumes:
      - mysql-data:/var/lib/mysql
      - ./mysql-init:/docker-entrypoint-initdb.d

  redis:
    image: redis:latest
    environment:
      - CELERY_BROKER_URL=redis://redis:6379/0
      - CELERY_RESULT_BACKEND=redis://redis:6379/0
      - SQLALCHEMY_DATABASE_URI=mysql+pymysql://root:root@mysql/fyp
      - SQLALCHEMY_TRACK_MODIFICATIONS=False
    ports:
      - "6380:6379"

volumes:
  mysql-data:
